pymeasure/instruments/agilent/agilent33500.py

```diff
@@ -0,0 +1,462 @@
+
+import logging
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import strict_discrete_set,\
+    strict_range, joined_validators
+from time import time
+from pyvisa.errors import VisaIOError
+
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+# Capitalize string arguments to allow for better conformity with other WFG's
+# FIXME: Currently not used since it does not combine well with the strict_discrete_set validator
+# def capitalize_string(string: str, *args, **kwargs):
+#     return string.upper()
+
+
+# Combine the capitalize function and validator
+# FIXME: This validator is not doing anything other then self.capitalize_string
+# FIXME: I removed it from this class for now
+# string_validator = joined_validators(capitalize_string, strict_discrete_set)
+
+
+class Agilent33500(Instrument):
+    """Represents the Agilent 33500 Function/Arbitrary Waveform Generator family.
+    Individual devices are represented by subclasses.
+
+    .. code-block:: python
+
+        generator = Agilent33500("GPIB::1")
+
+        generator.shape = 'SIN'                 # Sets the output signal shape to sine
+        generator.frequency = 1e3               # Sets the output frequency to 1 kHz
+        generator.amplitude = 5                 # Sets the output amplitude to 5 Vpp
+        generator.output = 'on'                 # Enables the output
+
+    """
+
+    # Validation ranges - These can be overridden by the child class to implement different validation
+    # ranges for different devices from the Agilent 33500 family.
+    # They are named using <method>_validation_range.
+
+    frequency_validation_range = [1e-6, 120e+6]
+    arb_srate_validation_range = [1e-6, 250e6]
+
+    id = Instrument.measurement(
+        "*IDN?", """ Reads the instrument identification """
+    )
+
+    def __init__(self, adapter, **kwargs):
+        super(Agilent33500, self).__init__(
+            adapter,
+            "Agilent 33500 Function/Arbitrary Waveform generator family",
+            **kwargs
+        )
+
+    def beep(self):
+        """ Causes a system beep. """
+        self.write("SYST:BEEP")
+
+    shape = Instrument.control(
+        "FUNC?", "FUNC %s",
+        """ A string property that controls the output waveform. Can be set to:
+        SIN<USOID>, SQU<ARE>, TRI<ANGLE>, RAMP, PULS<E>, PRBS,  NOIS<E>, ARB, DC. """,
+        validator=strict_discrete_set,
+        values=["SIN", "SQU", "TRI", "RAMP", "PULS", "PRBS", "NOIS", "ARB", "DC"],
+    )
+
+    frequency = Instrument.control(
+        "FREQ?", "FREQ %f",
+        """ A floating point property that controls the frequency of the output
+        waveform in Hz, from 1 uHz to 120 MHz (maximum range, can be lower depending
+        on your device), depending on the specified function. Can be set. """,
+        validator=strict_range,
+        values=frequency_validation_range,
```

Comment:
Can you confirm that the `frequency` control has the proper validation values for the Agilent33521A (child class)? I don't think it will actually get set properly because the `frequency` control will reference the values in the parent class. I think the simple solution is to overwrite this control in the child class. The same should go for `arb_srate`. Does that make sense?

Reply:
Oh, that's a tricky consequence of using these property factory in the class definition phase (in contrast to during instantiation).
I played with this a little, redefining the affected properties, although not too elegant, is probably the best way to do this (effort-wise), although I would love to be proven wrong!

Reply:
Good catch, it did not work. I agree, that for now, the best solution is the redefinition. I remove the variables, since they do not serve a purpose any more (ranges are defined in the redefined methods). A plus side if this is that the documentation now reflects the correct ranges in the child classes.

Reply:
Agreed. Thanks!